compiler/noirc_frontend/src/hir/def_collector/dc_mod.rs

```diff
@@ -684,26 +682,48 @@ impl<'a> ModCollector<'a> {
 fn find_module(
     file_manager: &FileManager,
     anchor: FileId,
-    mod_name: &str,
-) -> Result<FileId, String> {
+    mod_name: &Ident,
+) -> Result<FileId, DefCollectorErrorKind> {
     let anchor_path = file_manager
         .path(anchor)
         .expect("File must exist in file manager in order for us to be resolving its imports.")
         .with_extension("");
     let anchor_dir = anchor_path.parent().unwrap();
 
-    // if `anchor` is a `main.nr`, `lib.nr`, `mod.nr` or `{mod_name}.nr`, we check siblings of
-    // the anchor at `base/mod_name.nr`.
-    let candidate = if should_check_siblings_for_module(&anchor_path, anchor_dir) {
-        anchor_dir.join(format!("{mod_name}.{FILE_EXTENSION}"))
+    // Assuming anchor is called "anchor.nr" and we are looking up a module named "mod_name"...
+    // This is "mod_name"
+    let mod_name_str = &mod_name.0.contents;
+
+    // If we are in a special name like "main.nr", "lib.nr", "mod.nr" or "{mod_name}.nr",
+    // the search starts at the same directory, otherwise it starts in a nested directory.
+    let start_dir = if should_check_siblings_for_module(&anchor_path, anchor_dir) {
+        anchor_dir
     } else {
-        // Otherwise, we check for children of the anchor at `base/anchor/mod_name.nr`
-        anchor_path.join(format!("{mod_name}.{FILE_EXTENSION}"))
+        anchor_path.as_path()
     };
 
-    file_manager
-        .name_to_id(candidate.clone())
-        .ok_or_else(|| candidate.as_os_str().to_string_lossy().to_string())
+    // Check "mod_name.nr"
+    let mod_name_candidate = start_dir.join(format!("{mod_name_str}.{FILE_EXTENSION}"));
+    let mod_name_result = file_manager.name_to_id(mod_name_candidate.clone());
+
+    // Check "mod_name/mod.nr"
+    let mod_nr_candidate = start_dir.join(mod_name_str).join(format!("mod.{FILE_EXTENSION}"));
+    let mod_nr_result = file_manager.name_to_id(mod_nr_candidate.clone());
+
+    match (mod_nr_result, mod_name_result) {
+        (Some(_), Some(_)) => Err(DefCollectorErrorKind::OverlappingModuleDecls {
+            mod_name: mod_name.clone(),
+            expected_path: mod_name_candidate.as_os_str().to_string_lossy().to_string(),
+            alternative_path: mod_nr_candidate.as_os_str().to_string_lossy().to_string(),
+        }),
+        (Some(mod_nr_id), None) => Ok(mod_nr_id),
+        (None, Some(name_id)) => Ok(name_id),
```

Comment:
We can merge these two cases.
```suggestion
        (Some(id), None) | (None, Some(id)) => Ok(id),
```